### DISCUSSION TITLE
How to build a table schema with dynamic embedding model

### DISCUSSION THREAD

**Discussion by moehmeni** [lancedb/lancedb#1448] [Q&A]
I am reading Hybrid search docs and I should make the table like this:
```python
embeddings = get_registry().get("openai").create()

class Documents(LanceModel):
    vector: Vector(embeddings.ndims()) = embeddings.VectorField()
    text: str = embeddings.SourceField()

table = db.create_table("documents", schema=Documents)
```
what if I want to put this inside a function basd on embeddings that the user wants like:
```python
class LaunceDB:
    def __init__(self, path: str, embedding_model: str = "all-MiniLM-L6-v2"):
        self.db = lancedb.connect(path)
        embeddings = (
            get_registry().get("sentence-transformers").create(name=embedding_model)
        )
        ArticleScheme = # generate this
        self.table = self.db.create_table(
            "articles", schema=ArticleScheme, exist_ok=True
        )
        self.table.create_fts_index("text")  # for full-text search
        self.table.create_index()  # ANN for vector search
```
I want the schema to look like this:
```python
class ArticleScheme (LanceModel):
    vector: Vector(embeddings.ndims()) = embeddings.VectorField()
    text: str = embeddings.SourceField()
    db_id: int = None # I do not know if None is correct here, what should it be?
```
### Update:
I realized I can define classes in functions:
```python
class LaunceDB:
    def __init__(self, path: str, embedding_model: str = "all-MiniLM-L6-v2"):
        self.db = lancedb.connect(path)
        embeddings = (
           get_registry().get("sentence-transformers").create(name=embedding_model)
        )

        class ItemSchema(LanceModel):
            vector: Vector(embeddings.ndims()) = embeddings.VectorField()
            text: str = embeddings.SourceField()
            reference_id: int = Field(...)

        self.table = self.db.create_table("articles", schema=ItemSchema, exist_ok=True)
        self.table.create_fts_index("text")  # for full-text search
        self.table.create_index()  # ANN for vector search 
```
but now `vector: Vector(embeddings.ndims()) = embeddings.VectorField()` is giving me Pylance error:
Call expression not allowed in type expressionPylance[reportInvalidTypeForm](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportInvalidTypeForm)

**Comment by TayfurYldz** [comment]
The runtime part of your approach is valid; the Pylance diagnostic is about Python's **static annotation grammar**, not about LanceDB rejecting a dynamic vector dimension.

`lancedb.pydantic.Vector(dim)` is a runtime type factory. On current LanceDB it literally returns a generated `FixedSizeList` type whose `dim()` is the value passed to the factory. That is why this works at runtime inside a class body, but Pyright/Pylance reports a call expression used as a type expression.

If the schema genuinely depends on a model selected at runtime, I would make the model dynamic explicitly instead of suppressing the type checker. Pydantic's `create_model` works for this:

```python
from pydantic import Field, create_model
from lancedb.pydantic import LanceModel, Vector

embeddings = (
    get_registry()
    .get("sentence-transformers")
    .create(name=embedding_model)
)

ItemSchema = create_model(
    "ItemSchema",
    __base__=LanceModel,
    vector=(Vector(embeddings.ndims()), embeddings.VectorField()),
    text=(str, embeddings.SourceField()),
    reference_id=(int, Field(...)),
)

table = db.create_table("articles", schema=ItemSchema, exist_ok=True)
```

This preserves the two pieces LanceDB needs: the runtime `Vector(n)` type and the `VectorField`/`SourceField` metadata that records the embedding-function relationship.

For the other field, choose the annotation/default according to the actual schema contract:

```python
reference_id=(int, Field(...))          # required, non-null
# or
reference_id=(int | None, None)         # optional / nullable
```

`db_id: int = None` is not a good representation of the latter because the annotation says non-null `int` while the default says `None`.

So there are two reasonable cases: if the embedding model/dimension is fixed when you write the program, keep the normal declarative `class ...: vector: Vector(384)` form; if it is user-selected at runtime, build the Pydantic model at runtime as above. The latter also avoids asking a static type checker to understand a runtime-computed annotation.
